openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -408,24 +415,30 @@ def _refresh_addresses(self):
         handle cases in which the IP has changed
         """
         self.connection.device.refresh_from_db()
-        self.connection.refresh_from_db()
+        self.connection = self.connection.get_working_connection(self.connection.device)
         self.addresses = self.connection.get_addresses()
 
+    def _log_reconnecting_error(self, attempt):
+        addresses = ', '.join(self.addresses)
+        self.log(
+            _(
+                'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
+                    addresses=addresses, attempt=attempt
+                )
+            ),
+            save=False,
+        )
+
     def _write_checksum(self, checksum):
         for attempt in range(1, self.RECONNECT_MAX_RETRIES + 1):
-            self._refresh_addresses()
-            addresses = ', '.join(self.addresses)
-            self.log(
-                _(
-                    'Trying to reconnect to device at {addresses} (attempt n.{attempt})...'.format(
-                        addresses=addresses, attempt=attempt
-                    )
-                ),
-                save=False,
-            )
             try:
-                self.connect()
```

Comment:
We do not need to perform the connect operation here. `self.connection.get_working_connection` will open the SSH connection.